cpp/src/lists/extract.cu

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/gather.cuh>
+#include <cudf/lists/extract.hpp>
+
+#include <thrust/transform.h>
+
+namespace cudf {
+namespace lists {
+namespace detail {
+
+namespace {
+
+/**
+ * @brief Convert index value for each sublist into a gather index for
+ * the lists column's child column.
+ */
+template <bool PositiveIndex = true>
+struct map_index_fn {
+  column_device_view const d_offsets;  // offsets to each sublist (including validity mask)
+  size_type const index;               // index of element within each sublist
+  size_type const out_of_bounds;       // value to use to indicate out-of-bounds
+
+  __device__ int32_t operator()(size_type idx)
+  {
+    if (d_offsets.is_null(idx)) return out_of_bounds;
+    auto const offset = d_offsets.element<int32_t>(idx);
+    auto const length = d_offsets.element<int32_t>(idx + 1) - offset;
+    if (PositiveIndex)
+      return index < length ? index + offset : out_of_bounds;
+    else
+      return index < -length ? out_of_bounds : length + index + offset;
```

Comment:
I'd prefer this to be consistent with the previous line, i.e. valid ? compute : OOB
```suggestion
      return index >= -length ? length + index + offset : out_of_bounds;
```